### ISSUE TITLE
Solution different from Testresult in friendly-pow

### ISSUE THREAD

**Issue by gringofe** [FriendlyCaptcha/friendly-challenge#14]
Hello,

when I launch the `puzzle.jstest.ts` `generatePuzzleBuffer` in friendly-pow with the following parameters:

```
accountId: 0,
appId: 0,
numberOfPuzzles: 1,
puzzleExpiry: 100,
puzzleDifficulty: 30,
eightByteNonce: new Uint8Array(8),
nowInSeconds: 1610102754
```
the `puzzleFixture` value is ` Uint8Array [
      95,
      248,
      55,
      226,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      1,
      100,
      1,
      30,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0 ]`. The generated solution stored in `solutionsFixture` is ` Uint8Array [ 0, 0, 0, 0, 10, 0, 0, 0 ]`, which is correct when checked with `checkSolution`; But when I use the widget to solve this puzzle, it returns `Uint8Array(8) [ 0, 0, 0, 0, 2, 0, 0, 0 ]` as solution. Is there anything I'm not getting? Shouldn't these be the same?

**Comment by gzuidhof** [comment]
Hi gringofe,

That's interesting, are both valid solutions? I would expect both to be the same indeed as the attempts are made in the same order (starting from [0,0,0,0] going up until a valid solution is found). Was the widget solution also valid? Are you sure the inputs are the same? 

The relevant worker code can be found [here](https://github.com/FriendlyCaptcha/friendly-challenge/blob/master/src/worker.ts#L71-L89).

The widget directly imports the exact same solver code so it's hard to explain this. Note that both are instant (with the solvers solving hundreds of thousands to millions of attempts per second), but it may be looking into if this also applies to much higher difficulty levels.

**Comment by gringofe** [comment]
Hi,

no, only the solution generated by the backend code is valid. I also tried using the puzzle buffer your demo page receives from the backend (I decoded it from base64) and the solution the widget sent back to the demo server (also decoded that ofcourse), and passed it to [the solution method](https://github.com/FriendlyCaptcha/friendly-pow/blob/8bf8b6d0054c52e2313c66218c2f1fc894fb952b/src/verification.ts#L8) and it came back as invalid, even though it worked perfectly on the demo page.

**Comment by gzuidhof** [comment]
I see, that's surprising. `friendly-pow`'s package hasn't changed in many months now (latest version 0.1.2).

I had one theory: the version on NPM wasn't actually 0.1.2 (NPM doesn't actually check that):  I checked whether the generated wasm was any different between the latest version and 0.1.2 on NPM (in case a previous build was pushed to NPM), but it turned out to be the same.. So that's not it.

It could be that the JS and WASM version are different (the widget in your browser uses the latter probably), but our server uses the JS build too so it should have been caught there. The Friendly Captcha backend code uses 0.1.0, but the only difference between that and 0.1.2 is a small fix in the WASM version so it shouldn't matter. I'll bump it someday.

Maybe the WASM version is more strict? That has not been verified yet..

----
Looking into it a bit more, when running with the setup you provided:

```
   const difficulty = 30;
    const puzzleFixture = generatePuzzleBuffer({
        accountId: 0,
        appId: 0,
        numberOfPuzzles: 1,
        puzzleExpiry: 100,
        puzzleDifficulty: difficulty,
        eightByteNonce: new Uint8Array(8),
        nowInSeconds: 1610102754
        // userDataBuffer: new Uint8Array([1,2,3,4,5]),
    });

    const input = new Uint8Array(128);
    input.set(puzzleFixture.slice(), 0);
    solveBlake2bEfficient(input, difficultyToThreshold(difficulty), Math.pow(2, 32));
    const solutionsFixture = input.slice(120);
    console.log(solutionsFixture);
```
Running with `npm run test:ts` it prints
```
Uint8Array(8) [
        0, 0, 0, 0,
        2, 0, 0, 0
      ]
```
So that's different from your finding (and the same as the widget) which I can't explain. Printing the puzzleFixture yields `Uint8Array(32) [
        95, 248, 55, 226,   0, 0,  0, 0, 0,
         0,   0,  0,   1, 100, 1, 30, 0, 0,
         0,   0,  0,   0,   0, 0,  0, 0, 0,
         0,   0,  0,   0,   0
      ]` which looks the same.

When running with difficulty 31 instead, it needs 14 attempts. These are the hash values and the threshold (see the line number, it prints both sides of the condition in `src/solver.ts`).

```
Console

    console.log
      3001569514 292705578

      at Object.solveBlake2bEfficient (src/solver.ts:43:17)

    console.log
      871012196 292705578

      at Object.solveBlake2bEfficient (src/solver.ts:43:17)

    console.log
      769736140 292705578

      at Object.solveBlake2bEfficient (src/solver.ts:43:17)

    console.log
      306426772 292705578

      at Object.solveBlake2bEfficient (src/solver.ts:43:17)

    console.log
      352881499 292705578

      at Object.solveBlake2bEfficient (src/solver.ts:43:17)

    console.log
      3550837409 292705578

      at Object.solveBlake2bEfficient (src/solver.ts:43:17)

    console.log
      3165892701 292705578

      at Object.solveBlake2bEfficient (src/solver.ts:43:17)

    console.log
      1549355853 292705578

      at Object.solveBlake2bEfficient (src/solver.ts:43:17)

    console.log
      3992264059 292705578

      at Object.solveBlake2bEfficient (src/solver.ts:43:17)

    console.log
      1036761824 292705578

      at Object.solveBlake2bEfficient (src/solver.ts:43:17)

    console.log
      815307138 292705578

      at Object.solveBlake2bEfficient (src/solver.ts:43:17)

    console.log
      2483352516 292705578

      at Object.solveBlake2bEfficient (src/solver.ts:43:17)

    console.log
      1101127452 292705578

      at Object.solveBlake2bEfficient (src/solver.ts:43:17)

    console.log
      4623633 292705578

      at Object.solveBlake2bEfficient (src/solver.ts:43:17)

    console.log
      Uint8Array(8) [
         0, 0, 0, 0,
        13, 0, 0, 0
      ]

      at Suite.<anonymous> (test/puzzle.jstest.ts:22:13)
```
Do you see the same numbers?

**Comment by gringofe** [comment]
I just recloned the repo and it worked there, so no issue on your side. I then went to hunt for the issue and found nothing, which prompted me to run a diff between the two versions of `blake2b.ts`. This is what it outputted:

```
<       0xF3BCC908, 0x6A09E667, 0x84CAA73B, 0xBB67AE85,
---
>       0xF3BCC32908, 0x6A09E667, 0x84CAA73B, 0xBB67AE85,
```

Looks like i accidentially added `32` to one of the values in the intialization vector `BLAKE2B_IV32`.

God I feel stupid now, and am terribly sorry for wasting your time.
